### ISSUE TITLE
File download and list API for Playwright

### ISSUE THREAD

**Issue by vania-pooh** [aerokube/moon#417]
Need something similar to `/download` API in Selenium.

**Comment by facusantillo** [comment]
This is critical for us too, since we are moving to playwright from Selenium C# and we don't have a way to recover the files from the pod.

**Comment by keef78** [comment]
we need it as well, much needed to download files on playwright running on moon.  download to f.e. /home/user/Downloads doesnt work  (using robot framework with browser library (playwright)

**Comment by vania-pooh** [comment]
```
const { chromium } = require('playwright');
  
(async () => {
  var browser = await chromium.connect({ timeout: 0, wsEndpoint: 'ws://moon.aerokube.local/playwright/chrome/playwright-1.46.1?headless=false' });
  const page = await browser.newPage();
  await page.goto('https://kernel.org/');

  const downloadPromise = page.waitForEvent('download');
  await page.getByRole('link', { name: '6.10.8' }).click();
  const download = await downloadPromise;

  await download.saveAs(download.suggestedFilename());
  await browser.close();
})();
```

https://playwright.dev/docs/downloads

**Comment by keef78** [comment]
I am not sure I understand the comment, basically our issue is not with the download it self - file is downloaded - but it cannot be recovered from the "save as" lcoation - the lcoation is not available to the current running browser session - so in the case for example you download the PDF file as per documentation to  "/home/user/Downloads/some.pdf"    -   https://aerokube.com/moon/latest/#accessing-downloaded-files    , the session started by playwright cannot reach this destination. I have verified the  access rights are ok..

but when we try to serve the files in advance to the browser session  https://aerokube.com/moon/latest/#selenium-context - so during the start of the session browser is informed where are some file fot the test usage, like in /home/user/some.pdf  , the browser session in playwright on Moon can open this file ..

**Comment by facusantillo** [comment]
Hi,

What’s the status of this functionality? Our implementation is being delayed because we need this.

**Comment by aandryashin** [comment]
Will not be done. Playwright has its own functionality to work with files.

пт, 31 янв. 2025 г., 12:48 Facundo Santillo Alarcon <
***@***.***>:

**Comment by facusantillo** [comment]
Can you point then how can I recover files from the container itself?

**Comment by aandryashin** [comment]
const { chromium } = require('playwright');

(async () => {
  var browser = await chromium.connect({ timeout: 0, wsEndpoint:
'ws://moon.aerokube.local/playwright/chrome/playwright-1.46.1?headless=false'
});
  const page = await browser.newPage();
  await page.goto('https://kernel.org/');

  const downloadPromise = page.waitForEvent('download');
  await page.getByRole('link', { name: '6.10.8' }).click();
  const download = await downloadPromise;

  await download.saveAs(download.suggestedFilename());
  await browser.close();
})();

пт, 31 янв. 2025 г., 20:38 Facundo Santillo Alarcon <
***@***.***>:

**Comment by facusantillo** [comment]
I'll give it a try and come back with an answer, but last time I checked the file was downloaded inside the container and it was impossible to recover it from there.

I know this script will download the file, what I'm asking is how to access it later to check the file

**Comment by keef78** [comment]
that is exactly same issue in our environment, using licenced AeroKube / MOON as I stated above. 

Download works fine  BUT  we cannot access "/home/user/Downloads"  (no issues with priviledges at all) - to retrieve file and open it in browser with in SAME browser context - seems the file is simply not downloaded there.  We tried to use curl to list current files downloaded to path above, also no luck.

when downloaded to workspace of the running jenkins job, like SaveAS=my output directory , its downloaed fine and file exists - BUT - that is outside of browser container  - hence cannot be opened after download within the same browser, which is pretty annoying for tests requiring such flow.

We can ONLY send file into NEW browser context (using wss arguments like )

${MOON_URL}/playwright/chrome/playwright-${PLAYWRIGHT_VERSION}?
                     headless=${HEADLESS}
                     &enableVideo=true
                     &name=${BUILDTAG_ENV}
                     &arg=--proxy-pac-url=${PROXY_WPAD}
                     &${BROWSER_DATA}`

then we can access  fles sent to container  in  /home/user/filename BUT we dont need that , we need to be able to use files downloaded into current container


so to recap, we can download outside container but not able to open that file in the same browser session. when downloaded into the browsers container to /home/user/Downloads , it cannot be accessed either as the file is not there.

so we must use work arounds for scenario - download file - open it in the same browser , like download to workspace, grep what you need by custom scripts as its accessible of yourse in my working directory, continue with tests in the same browser session.

**Comment by aandryashin** [comment]
Playwright runs the browser in incognito mode, that is why you cannot
access downloaded files within container.

пт, 31 янв. 2025 г. в 21:49, keef78 ***@***.***>:
